jax/core.py

```diff
@@ -2243,51 +2283,73 @@ def ctx_factory():
 
 def _check_jaxpr(
     ctx_factory: Callable[[], Tuple[JaxprPpContext, JaxprPpSettings]],
-    jaxpr: Jaxpr,
-    in_avals: Sequence[AbstractValue]) -> None:
-
-  def read(v: Atom) -> AbstractValue:
-    if isinstance(v, Literal):
-      return raise_to_shaped(get_aval(v.val))
-    else:
-      if v not in env:
+    jaxpr: Jaxpr
+  ) -> None:
+  # Use set of variables to types to check that variables are in scope.
+  env: Set[Var] = set()
+
+  def read(x: Atom) -> Atom:
+    # Check the type annotation is itself well-typed.
+    check_type(ctx_factory, env, x.aval)
+    if isinstance(x, Var):
+      # Check the variable is in-scope and consistently typed.
+      if x not in env:
+        ctx, _ = ctx_factory()
+        raise JaxprTypeError(f"Variable '{pp_var(x, ctx)}' not defined")
+      return x
+    elif isinstance(x, Literal):
+      # Check that the literal matches its type annotation.
+      if not typecheck(x.aval, x.val):
         ctx, _ = ctx_factory()
-        raise JaxprTypeError(f"Variable '{pp_var(v, ctx)}' not defined")
-      return env[v]
+        raise JaxprTypeError(
+            f"Literal value {x.val} does not match its type annotation "
+            f"{pp_aval(x.aval, ctx)}")
+      return x
+    else:
+      assert False, "syntactically invalid jaxpr"
```

Comment:
```suggestion
      raise JaxprTypeError(f"syntactically invalid jaxpr: attempt to read {x}")
```

Reply:
I wanted to distinguish jaxpr syntax errors from jaxpr type errors, and assume `check_jaxpr` is only given syntactically valid jaxprs (hence the assert). Same below. WDYT?

Reply:
Put another way, we can and should have error checks for these in the `__init__` methods of the relevant classes (e.g. for this line the corresponding check would be in `JaxprEqn.__init__` asserting that each element of `invars` is an `Atom`).

Reply:
I see. Yes, I agree it makes sense to distinguish.

With that in mind, should we check for syntactic validity at any point as well? If so, would we do syntax checks under `check_jaxpr`, and raise `JaxprSyntaxError` rather than `JaxprTypeError`?

This question of whether to syntax-check, and where, is broader in scope than this PR, so we can proceed with `assert` for now and return to it.

Reply:
Well, if we have the checks in the `__init__` methods, then it won't be possible to construct a syntactically invalid jaxpr (without mutation), so we may not need a syntax checker.

Reply:
Sure, that'd work.